clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsGroupHeartbeatRequestManager.java

```diff
@@ -0,0 +1,387 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals;
+
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.internals.events.BackgroundEventHandler;
+import org.apache.kafka.clients.consumer.internals.metrics.HeartbeatMetricsManager;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.message.StreamsGroupHeartbeatRequestData;
+import org.apache.kafka.common.message.StreamsGroupHeartbeatResponseData;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.StreamsGroupHeartbeatRequest;
+import org.apache.kafka.common.requests.StreamsGroupHeartbeatResponse;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Timer;
+
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+
+public class StreamsGroupHeartbeatRequestManager implements RequestManager {
+
+    static class HeartbeatState {
+
+        private final StreamsMembershipManager membershipManager;
+        private final int rebalanceTimeoutMs;
+
+        private final StreamsRebalanceData streamsRebalanceData;
+
+        public HeartbeatState(final StreamsRebalanceData streamsRebalanceData,
+                              final StreamsMembershipManager membershipManager,
+                              final int rebalanceTimeoutMs) {
+
+            this.membershipManager = membershipManager;
+            this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+            this.streamsRebalanceData = streamsRebalanceData;
+        }
+
+        public StreamsGroupHeartbeatRequestData buildRequestData() {
+            StreamsGroupHeartbeatRequestData data = new StreamsGroupHeartbeatRequestData();
+            data.setGroupId(membershipManager.groupId());
+            data.setMemberId(membershipManager.memberId());
+            data.setMemberEpoch(membershipManager.memberEpoch());
+            membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+            StreamsGroupHeartbeatRequestData.Topology topology = new StreamsGroupHeartbeatRequestData.Topology();
+            topology.setSubtopologies(getTopologyFromStreams());
+            topology.setEpoch(streamsRebalanceData.topologyEpoch());
+            data.setRebalanceTimeoutMs(rebalanceTimeoutMs);
+            data.setTopology(topology);
+            data.setProcessId(streamsRebalanceData.processId().toString());
+            data.setActiveTasks(Collections.emptyList());
```

Comment:
Why do you se this to empty list and overwrite it a few lines down?

Reply:
That was an oversight that originates from reducing the size of the PR. Thanks!